enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,947 @@
+used for allocating pod IPs on the cluster default network today.
+* The cluster default primary network defined today will remain in place as the default network pods attach to. The cluster
+default network will continue to serve as the primary network for pods in a namespace that has no user-defined network. Pods
+with user-defined networks will still attach to the cluster default network with limited access to Kubernetes system resources.
+Pods with user-defined networks will have at least two network interfaces, one connected to the cluster default network and one
+connected to the user-defined network. Pods with user-defined networks will use the user-defined network as their default
+gateway.
+* Allow multiple namespaces per network.
+* Support cluster ingress/egress traffic for user-defined networks, including secondary networks.
+* Support for ingress/egress features where possible:
+    * EgressQoS
+    * EgressService
+    * EgressIP
+    * Load Balancer and NodePort Services, as well as services with External IPs.
+* In addition to ingress service support, there will be support for Kubernetes services in user-defined networks. The
+scope of reachability to that service as well as endpoints selected for that service will be confined to the network
+and corresponding namespace(s) where that service was created.
+* Support for pods to continue to have access to the cluster default primary network for DNS and KAPI service access.
+* Kubelet healthchecks/probes will still work on all pods.
+* OpenShift Router/Ingress will work with some limitations for user-defined networks.
+
+### Non-Goals
+
+* Allowing different service CIDRs to be used in different networks.
+* Localnet will not be supported initially for primary networks.
+* Allowing multiple primary networks per namespace.
+* Hybrid overlay support on user-defined networks.
+
+### Future-Goals
+
+* DNS lookup for pods returning records for IPs on the user-defined network. In the first phase DNS will return the pod
+IP on the cluster default network instead.
+* Admin ability to configure networks to have access to all services and/or expose services to be accessible from all
+networks.
+* Ability to advertise user-defined networks to external networks using BGP/EVPN. This will enable things like:
+    * External -> Pod ingress per VRF (Ingress directly to pod IP)
+    * Multiple External Gateway (MEG) in a BGP context, with ECMP routes
+* Allow connection of multiple networks via explicit router API configuration.
+
+## Proposal
+
+By default in OVN-Kubernetes, pods are attached to what is known as the “cluster default" network, which is a routed network
+divided up into a subnet per node. All pods will continue to have an attachment to this network, even when assigned a
+different primary network. Therefore, when a pod is assigned to a user-defined network, it will have two interfaces, one
```

Comment:
Maybe we can have just one interface at the pods if for example in the case of layer3 primary network we connect its ovn_cluster_router and the default network ovn_cluster_router together using the transis switches.

We are already connecting both networks but using pods instead of ovn.

For sure I am missing a lot of stuff.

Reply:
yeah it's a possibility, but then we would have to implement a bunch of access-control inside of OVN to not allow routing to the other networks via the default ovn_cluster_router. Additionally it becomes hard to implement host to pod access. Since the pods would have kube primary IP as their unique IP entry. They could have overlapping IPs. Then when kubelet goes to probe or access one, we wouldn't know how to get to the "right" pod.

So then the next logical step is OK, then put default network IP and user-defined IP on the same pod interface, and keep the pod primary IP address as the cluster default one. But at that point its not really worth all the new OVN topology changes we would have to do and easier to just plug in another NIC to the pod for cluster default network access.

Reply:
make sense, also I suppose it collide with stuff like SRIOV.

Thanks for explanation.